File: androidApp/src/main/java/com/dhruv/myfirstkmm/android/note_list/NoteListViewModel.kt
package com.dhruv.myfirstkmm.android.note_list

import android.database.DatabaseUtils
import android.graphics.Color
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dhruv.myfirstkmm.Data.note.SearchNote
import com.dhruv.myfirstkmm.Domain.note.Note
import com.dhruv.myfirstkmm.Domain.note.NoteDataSource
import com.dhruv.myfirstkmm.Domain.time.DateTimeUtil
import com.dhruv.myfirstkmm.Presentation.BabyBlueHex
import com.dhruv.myfirstkmm.Presentation.LightGreenHex
import com.dhruv.myfirstkmm.Presentation.RedOrangeHex
import com.dhruv.myfirstkmm.Presentation.RedPinkHex
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.LocalDateTime
import kotlin.random.Random

import javax.inject.Inject

@HiltViewModel
class NoteListViewModel @Inject constructor(
    private val noteDataSource : NoteDataSource,
    private val saveStateHandel : SavedStateHandle
) : ViewModel() {

    private val searchNotes = SearchNote()


    init {
        viewModelScope.launch {
             (1..10).forEach {
                noteDataSource.insertNote(
                    Note(
                        id = it.toLong(),
                        title = "Note $it",
                        content = "My note is good $it",
                        colorHex = BabyBlueHex,
                        created = DateTimeUtil.now()
                    )
                )
             }
        }
    }

    private val notes = saveStateHandel.getStateFlow("notes", emptyList<Note>())
    private val searchText = saveStateHandel.getStateFlow("searchText","")
    private val isSearchActive = saveStateHandel.getStateFlow("isSearchActive",false)

    val state = combine(notes,searchText,isSearchActive){ notes, seachText, isSearchText ->
        NoteListState(
            noteList = searchNotes.execute(notes,seachText),
            searchText = seachText,
            isSearchFieldActive = isSearchActive.value
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(), NoteListState())

    fun loadNotes(){
        viewModelScope.launch {
            saveStateHandel["notes"] = noteDataSource.getAllNotes()
        }
    }

    fun onSeachTextChange(text: String){
        saveStateHandel["searchText"] = text
    }

    fun onToggleSearch(){
        saveStateHandel["isSearchActive"] = !isSearchActive.value
        if (!isSearchActive.value){
            saveStateHandel["searchText"] = ""
        }
    }

    fun onDeleteNoteById(id: Long){
        viewModelScope.launch {
            noteDataSource.deleteNoteById(id)
            loadNotes()
        }
    }
}

File: iosApp/iosApp/ note_list/NoteListViewModel.swift
//
//  NoteListViewModel.swift
//  iosApp
//
//  Created by Dhruv Sharma on 04/04/23.
//  Copyright © 2023 Dhruv🐵. All rights reserved.
//

import Foundation
import shared

extension NoteListScreen{
    
    @MainActor class NoteListViewModel : ObservableObject{
        
        private let searchNotes = SearchNotes()
        
        private var notes = [Note]()
        @Published private(set) var filteredNotes = [Note]()
        
        @Published var searchText = " " {
            didSet{
                self.filteredNotes = searchNotes.execute(notes: self.notes,query: searchText)
            }
        }
        
        @Published private(set) var isSearchActive = false
 
        private var notedataSource: NoteDataSource? = nil
        
        init(notedataSource: NoteDataSource? = nil) {
            self.notedataSource = notedataSource
        }
        
        func setNoteDataSource(noteDataSource : NoteDataSource){
            self.notedataSource = noteDataSource
//            noteDataSource.insertNote(note: Note(id: nil, title: "First Note", content: "This is content of first note", colorHex: 0xff4434, created: DateTimeUtil().now())) { error in
//
//            }
        }
        
        
        func loadNotes(){
            notedataSource?.getAllNotes(completionHandler: { notes,error in
                self.notes = notes ?? []
                self.filteredNotes = self.notes
            })
        }
        
        func deleteNoteById(id: Int64?){
            if id != nil{
                notedataSource?.deleteNoteById(id: id!, completionHandler: { Error in
                    self.loadNotes()
                })
            }
        }
        
        func toggleIsSearchActive(){
            isSearchActive = !isSearchActive
            if !isSearchActive {  /// If not focus reset `searchText`
                searchText = ""
            }
            
        }
    }
}
